hazelcast/src/main/java/com/hazelcast/map/impl/operation/steps/engine/StepSupplier.java

```diff
@@ -59,17 +63,58 @@ public StepSupplier(MapOperation operation, boolean checkCurrentThread) {
         assert operation != null;
 
         this.state = operation.createState();
-        Step injectedStep = operation.getRecordStore().getStorage().newInjectedStep();
+        Step injectedStep = getInjectedCompositeStep(operation);
         Step currentFirstStep = operation.getStartingStep();
         this.currentStep = injectedStep == null
-                ? currentFirstStep : setAndGetInjectedStepAsFirstStep(currentFirstStep, injectedStep);
+            ? currentFirstStep : setAndGetInjectedStepAsFirstStep(currentFirstStep, injectedStep);
         this.operationRunner = UtilSteps.getPartitionOperationRunner(state);
         this.checkCurrentThread = checkCurrentThread;
 
         assert state != null;
         assert this.currentStep != null;
     }
 
+    /**
+     * Creates a composite step to perform compaction of record store
+     * as well as local indexes storage. If there is no local indexes on record
+     * store, composite Step is not used.
+     * @param operation the map operation
+     * @return the injected Step
+     */
+    private Step getInjectedCompositeStep(MapOperation operation) {
+        Storage storage = operation.getRecordStore().getStorage();
+
+        if (!(storage instanceof StepAwareStorage)) {
+            return null;
+        }
+
+        StepAwareStorage stepAwareStorage = (StepAwareStorage) storage;
+        Step primaryInjectedStep = stepAwareStorage.newInjectedStep();
+        if (operation.getMapContainer().isGlobalIndexEnabled()) {
```

Comment:
can we keep this StepSupplier class isolated as much as possible to prevent leaking of other concerns in it?

I suggest to have a separate PR for the StepSupplier changes. 
And proposing to have a change like this to add multiple Steps.
How do you see this?

```suggestion
public StepSupplier(MapOperation operation, boolean checkCurrentThread) {
        assert operation != null;

        this.state = operation.createState();
        List<Step> injectedSteps = operation.getRecordStore().getStorage().newInjectedSteps();
        Step currentFirstStep = operation.getStartingStep();
        this.currentStep = currentFirstStep;

        for (Step injectedStep : injectedSteps) {
            this.currentStep = setAndGetInjectedStepAsFirstStep(currentFirstStep, injectedStep);
            currentFirstStep = currentStep;
        }

        this.operationRunner = UtilSteps.getPartitionOperationRunner(state);
        this.checkCurrentThread = checkCurrentThread;

        assert state != null;
        assert this.currentStep != null;
    }
```

Reply:
@ahmetmircik I'd add this change in the follow up PR. Is it OK?

Reply:
based on our talk, created this: https://github.com/hazelcast/hazelcast/pull/25399